KarhooSDK/Network/HttpClient/TokenRefreshingHttpClient/Interactor/KarhooRefreshTokenInteractor.swift

```diff
@@ -140,10 +164,57 @@ final class KarhooRefreshTokenInteractor: RefreshTokenInteractor {
             return true
         }
         let timeToExpiration = date.timeIntervalSince1970 - Date().timeIntervalSince1970
-        return timeToExpiration < Constants.MaxTimeIntervalToRefreshToken
+        return timeToExpiration < Constants.refreshBuffer
     }
 
-    private struct Constants {
-        static let MaxTimeIntervalToRefreshToken = TimeInterval(30) // 30 sec
+    // MARK: Timers scheduling
+
+    private func scheduleRefreshTokenTimer() {
+        guard let credentials = dataStore.getCurrentCredentials() else {
+            assertionFailure("Credentials should be set at this stage")
+            return
+        }
+        invalidateRefreshTokenTimer()
+
+        let secondsToRefresh: TimeInterval = max(0, (credentials.expiryDate?.timeIntervalSinceNow ?? 0) - Constants.refreshBuffer)
+
+        refreshTokenTimer = Timer.scheduledTimer(
+            withTimeInterval: secondsToRefresh,
+            repeats: false,
+            block: { [weak self] _ in
+                self?.refreshToken { result in
+                    print("KarhooRefreshTokenInteractor.proactivelyRefreshToken result: \(result)")
+                }
+            }
+        )
+        RunLoop.main.add(refreshTokenTimer!, forMode: .common)
+    }
+
+    /// Schedule a timer that will return extenal authentition error if given time interval will be reached. This is designed as fallback in case of invalid DPs implementation of the `requireSDKAuthentication` method.
+    private func scheduleExtenalAuthInvalidationTimer() {
+        invalidateExternalAuthTimer()
+        externalAuthInvalidationTimer = Timer.scheduledTimer(
+            withTimeInterval: Constants.allowedExternalAuthTimeInterval,
+            repeats: false,
+            block: { [weak self] _ in
+                guard let self else { return }
+                self.invalidateExternalAuthTimer()
+                self.callback?(.failure(error: RefreshTokenError.extenalAuthenticationRequestExpired))
```

Comment:
Please make sure to send the same error as on Android to avoid confusion on debugging it if we get to this scenario with a DP.

Reply:
Aligned :+1: